### PR TITLE
Fix for A+B-->A|B breaking gather elimination in LLVM 11+

### PR DESCRIPTION
**Pull request by aschrein**
Fix for #1673

### CODE REVIEW

**Comment by aschrein** [comment]
Probably it makes sense to enable this patch for all LLVM versions as there's no test tasks for 11

File: src/util.cpp [hunk 1]
```diff
@@ -334,7 +334,7 @@ static void lPrint(const char *type, bool isError, SourcePos p, const char *fmt,
     // exact same error message.  If so, return, so we don't redundantly
     // print it and annoy the user.
     static std::set<std::string> printed;
-    if (printed.find(formattedBuf) != printed.end()) {
+    if (isError && printed.find(formattedBuf) != printed.end()) {
```
Comment (dbabokin) [inline src/util.cpp]: Why do need that?

We don't want to duplicate both errors and warnings.

**Comment by dbabokin** [comment]
The fix looks reasonable, but running it with trunk I don't see improvements. I.e. all build of examples is still producing warnings and "make check-all" yields 4 fails.

Can you also add a lit test?

**Comment by DeepakRajendrakumaran** [comment]
Some additional notes-

This and it’s following changes in LLVM is what seem to cause it - https://reviews.llvm.org/D73435

The patch fixes some cases. But there are others which still remains

The case below is an example where an add got converted to OR. Something like this is fixed

varying float
IntersectLightsWithTileMinMax(uniform float light_positionView_z_array[],uniform int32 numLights)
{
        float light_positionView_z;
       foreach (lightIndex = 0 ... numLights)
              light_positionView_z = light_positionView_z_array[lightIndex];

      return light_positionView_z;
}

File: src/util.cpp (hunk 1 above)
Comment (aschrein) [reply src/util.cpp]: This routine is used for debug output. I think we do need debug messages whether they're duplicate or not.

**Comment by aschrein** [comment]
> "make check-all" yields 4 fails

Those tests fail under LLVM 11 independent of this patch. No fails with LLVM 9.

> all build of examples is still producing warnings

I've been looking at this simple example:
```c++
```

File: src/util.cpp (hunk 1 above)
Comment (dbabokin) [reply src/util.cpp]: It's used for error, warnings (including performance warnings), and debug prints.

We definitely don't want duplication for errors and warnings. Debug prints should allow duplication.

The reason for duplicated errors and warnings is that we duplicate extern functions (masked and unmasked version) and they trigger the same messages, which is very annoying and confusing.

Comment (aschrein) [reply src/util.cpp]: Ok, I removed this change. It requires some extra efforts.

**Pull request merged**